src/resources.rs

```diff
@@ -107,34 +161,104 @@ fn assets_in_markdown(src: &str, parent_dir: &Path) -> Result<Vec<PathBuf>, Erro
 
     found.sort();
     found.dedup();
+    if !found.is_empty() {
+        trace!("Assets found in content : {:?}", found);
+    }
+    Ok(found)
+}
 
-    // TODO: Allow linked images to be either a URL or path on disk
+pub(crate) fn hash_link(url: &Url) -> String {
+    use std::collections::hash_map::DefaultHasher;
+    use std::hash::{Hash, Hasher};
 
-    // I'm assuming you'd just determine if each link is a URL or filename so
-    // the `find()` function can put together a deduplicated list of URLs and
-    // try to download all of them (in parallel?) to a temporary location. It'd
-    // be nice if we could have some sort of caching mechanism by using the
-    // destination directory (hash the URL and store it as
-    // `book/epub/cache/$hash.$ext`?).
-    let mut assets = Vec::new();
+    let mut hasher = DefaultHasher::new();
+    url.hash(&mut hasher);
+    let path = PathBuf::from(url.path());
+    let ext = path
+        .extension()
+        .and_then(OsStr::to_str)
+        .unwrap_or_else(|| panic!("Unable to extract file ext from {url}"));
+    format!("{:x}.{}", hasher.finish(), ext)
+}
 
-    for link in found {
-        let link = PathBuf::from(link);
-        let filename = parent_dir.join(link);
-        let filename = filename.canonicalize()?;
+// From cargo/util/paths.rs
+pub fn normalize_path(path: &Path) -> PathBuf {
+    let mut components = path.components().peekable();
+    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+        components.next();
+        PathBuf::from(c.as_os_str())
+    } else {
+        PathBuf::new()
+    };
 
-        if !filename.is_file() {
-            return Err(Error::AssetFile(filename));
+    for component in components {
+        match component {
+            Component::Prefix(..) => unreachable!(),
+            Component::RootDir => {
+                ret.push(component.as_os_str());
+            }
+            Component::CurDir => {}
+            Component::ParentDir => {
+                ret.pop();
+            }
+            Component::Normal(c) => {
+                ret.push(c);
+            }
         }
-
-        assets.push(filename);
     }
-    trace!("Assets found in content : [{}]", assets.len());
-    Ok(assets)
+    ret
 }
 
+pub(crate) mod handler {
+    use std::{
+        fs::{self, File, OpenOptions},
+        io::{self, Read},
+        path::Path,
+    };
+
+    #[cfg(test)]
+    use mockall::automock;
+
+    use crate::Error;
+
+    use super::{Asset, AssetKind};
+
+    #[cfg_attr(test, automock)]
+    pub(crate) trait ContentRetriever {
+        fn download(&self, assets: &Asset) -> Result<(), Error>;
+        fn read(&self, path: &Path, buffer: &mut Vec<u8>) -> Result<(), Error>;
+    }
+
+    pub(crate) struct ResourceHandler;
+
+    impl ContentRetriever for ResourceHandler {
+        fn download(&self, asset: &Asset) -> Result<(), Error> {
```

Comment:
As I can see `download(..)` is not covered by tests. Could you pls add error (missing remote resource) and happy paths for that ? Mocking remote will be fine I think.

Reply:
Using `AssetFileNotFound`  should be enough I think.